### PR TITLE
[Drivers] Add support for a Redis Driver

### PR DESCRIPTION
**Pull request by Drapersniper**
Add the Redis Driver.

NOTE: THIS WILL NOT WORK with a simple Redis driver, You will need the ReJSON plugin enabled in the Redis server.
More instructions can be found in https://rejson.io/
If you want a simple Docker image with it already built and enable use `redislabs/redisjson2:latest` otherwise to do it manually follow https://oss.redislabs.com/redisjson/#building-and-loading-the-module

**WARNING**: __If you are doing this you need to ensure your redis server is persisting data, otherwise all your data will be wiped in a system restart.__

# Setup
## Hosting the server with Docker

- Saved the following in a file called `docker-compose.yaml`

```yaml
version: "3.7"
services:
  redisconfig-helper:
    container_name: redisconfig-helper
    image: alpine
    command: sh -c "echo never > /sys/kernel/mm/transparent_hugepage/enabled && echo never > /sys/kernel/mm/transparent_hugepage/defrag"
    privileged: true
  redisconfig:
    container_name: redisconfig
    image: redislabs/rejson:latest
    restart: always
    privileged: true
    sysctls:
      net.core.somaxconn: "65535"
      net.ipv4.tcp_max_syn_backlog: "3240000"
      net.ipv4.tcp_wmem: "4096 65536 524288"
      net.ipv4.tcp_rmem: "4096 87380 524288"
    volumes:
      - /home/user/redisserver/config/redis.conf:/usr/local/etc/redis/redis.conf # This is your config file
      - /home/user/redisserver/data:/data # This is your Redis data folder
      - /home/user/redisserver/sock:/tmp # This is your Unix Socket folder, you will need to run `chown 999 /home/user/redisserver/sock` to change the perms of this folder before starting the docker
    networks:
      redisconfig-net:
        ipv4_address: 172.27.0.2
    ulimits:
      nofile:
        soft: 65536
        hard: 65536
    depends_on:
      - redisconfig-helper
    command: ["redis-server", "/usr/local/etc/redis/redis.conf"]

networks:
  redisconfig-net:
    name: redisconfig-net
    driver: bridge
    ipam:
      config:
        - subnet: 172.27.0.0/16
```

- Create a file: `<full path to redis storage>\redis.conf` and paste the content of this https://gist.github.com/Drapersniper/b119bf588c0890af75659373579568e2 config inside of it
 - By default the password for the server is `foorbarred`
 - To edit the password open `redis.conf` and edit the following line:
   - `requirepass foorbarred` to `requirepass <my new password>`

- Start the reJSON server: `cd` into the folder where you saved the `docker-compose.yaml` and run `docker-compose up -d` 

## Migrate from JSON to Redis
- run : `redbot-setup convert <instance_name> redis`
- Follow the prompts to connecto to your redis server. (Note the IP for the redis server will be `172.27.0.2`)

- Once migration is completed start the bot

### Edit 21/09/20
- Added the sock folder mount to the compose for users wishing to setup the server with unix sockets.
Note:
```config
unixsocket /tmp/redis.sock
unixsocketperm 777
```
add this to the end of your `redis.conf` file you've downloaded, if you wish to setup the redis server with UNIX sockets.

### CODE REVIEW

**Comment by Jackenmen** [comment]
I removed the "Blocked by: Other PR" label as #3714 is merged already but this PR now needs conflict resolution.

File: setup.cfg [hunk 1]
```diff
@@ -79,6 +79,9 @@ docs =
     urllib3==1.25.8
 postgres =
     asyncpg==0.20.1
+redis =
+    aioredis==1.3.1
```
Comment (Jackenmen) [inline setup.cfg:99]: This is missing `aioredis_lock` (though it seems like you're not using it at all)

File: setup.cfg [hunk 2]
```diff
@@ -79,6 +79,9 @@ docs =
...
+    ujson==2.0.3; sys_platform == "linux"
```
Comment (Jackenmen) [inline setup.cfg]: Primary requirements need to be added to `tools/primary_deps.ini`

**Comment by Jackenmen** [comment]
Info about the `redis` extra needs to be added to install_windows, install_linux_mac, and update_red documents in docs.

If you need help with setting up tests/CI, here's a commit adding those for redis, feel free to use it if you want to (fully working, CI fails only because you didn't add `aioredis_lock` to `redis` extra):
https://github.com/jack1142/Red-DiscordBot/commit/4b9f2d468585f970ab20b3bb5a83299cad9f5bcb

**Comment by Tobotimus** [comment]
Could we get redis tests done in CI on this PR? So we can easily track how it's going passing tests and such

**Comment by Drapersniper** [comment]
Jack has a commit for it somewhere I been sorta busy with other stuff but I'll add test prior to it being fully reviewed

**Comment by Jackenmen** [comment]
I fixed the stuff I mentioned in my review comments and added tox env for Redis and job in CI workflow.

Small reminder that when this gets merged, `Tests / Tox - Redis (3.8) (pull_request)` needs to be made a required check.

**Comment by mikeshardmind** [comment]
This PR in it of itself isn't breaking end users directly, but as discussed in discord, there needs to be some restrictions on config stored values, a breaking window, and tools for cog creators.

Limits needed by this PR:

64-bit integer limits and disallowing `NaN` and `inf` (positive and negative of each) in floats.

Further limit which should really be fine and may allow other improvements in the future from one set of restrictions, but that I don't think is strictly enforced currently: utf-8 strings. (all messages from discord are utf-8, python source code is utf-8, you have to go out of your way to get something else here)

This is probably going to need an issue and discussion of how to handle this separately.

While a 53 bit integer limit would maximize the potential for future library use, given that discord IDs are 64bit, this seems like a limit that can't be used without massive impact which is not worth the potential gains.

**Comment by PredaaA** [comment]
Posting this here so it doesn't get lost:
Currently there is an issue with the Redis driver, which is basically that a cog named `Space` (maybe other names too?) will *not* register the data, but only an empty dict.
Here is a screenshot of the issue in action:
image

**Review by laggron42 — DISMISSED** [review]
Been using Redis since almost a month, and it works super well! Everything is way faster, and surprisingly doesn't use that much RAM (less than my other big bot with a lot of cogs, 4.6% with 3.85GB of RAM).
It's working on a bot made for a 23k members server, with 38 cogs (including adventure and warnsystem, probably the most consuming cogs, and the difference between JSON and Redis can be clearly noticed by everyone for those commands, for example !loot took ~3 seconds, now it's almost instant).

**Comment by Stonedestroyer** [comment]
Can confirm Redis works well. 

Here is a systemd snippet which you have to modify incase you want Red to wait for redis. redis-cli is required on the server. 

```bash
[Service]
Environment="REDIS_PASSWORD=passwordhere"
ExecStart=/bin/bash -c ' \
  until [ "`redis-cli -h 127.0.0.1 -p porthere -a ${REDIS_PASSWORD} PING`" == "PONG" ]; do sleep 0.1; done; \
  /home/user/env/bin/python3 -O -m redbot instancename --no-prompt --dev'
```

**Comment by Stonedestroyer** [comment]
Months after using this I can confirm this is super smooth and made my bot much faster and allows me to code less efficient now cause config will take the load.

**Comment by kablekompany** [comment]
Made the change to using this for my hosted Red. 
Truly night and day difference in pretty well every aspect as far as response time, and performance re: system and connectivity. The provided walkthru was followed and on completing was success. A few things to note in the documentation would be validating updated installs of docker/docker-compose and defining your own paths for the docker-compose.yaml file to launch your container. 

Dope PR

**Comment by TrustyJAID** [comment]
I have tested this for over a week and had zero issues with it. It took some figuring out for proper wait for redis initialization in systemd but we figured it out here is an example of what worked for me.
```
#! /usr/bin/env bash
. /home/trusty/.redis_pass.sh
until redis-cli -h $REDIS_IP -a $REDIS_PASSWORD -s $REDIS_UNIX_SOCKET PING | grep "PONG";
do
  clear;
  echo "Waiting for Redis...";
  sleep 0.1;
done
```
```
[Unit]
Description=%I redbot
After=multi-user.target

[Service]
Environment="REDIS_PASSWORD=$REDIS_PASSWORD"
User=trusty
Group=trusty
Type=idle
Restart=always
RestartSec=15
RestartPreventExitStatus=0
TimeoutStopSec=10
LimitNOFILE=500000
ExecStart=/home/trusty/redredis/bin/redbot %I --no-prompt --rpc --enable-sudo --message-cache-size 5000
ExecPreStart=/bin/bash/ /home/trusty/redis_wait.sh
[Install]
WantedBy=multi-user.target
```
Technically speaking I couldn't see any *real* difference between this and previously used json driver. However, anecdotally I have noticed that there's a slight latency when using this compared to the json driver. That's not to say there isn't a need for this in some situations and that we should be eeking out the maximum performance latencies from our bots but it is just an observation I wanted to make note of here since I will be switching back to json driver for the time being.

**Comment by laggron42** [comment]
While trying to backup my instance using `redbot-setup backup`, I got the following error:
```py
redbot@laggron-VPS:~$ redbot-setup backup ATOS
[2021-02-25 15:08:51] [WARNING] red.config: DANGER! This is getting an outdated driver. Hopefully this is only being done from convert
Backing up the instance's data...
Traceback (most recent call last):
  File "/home/redbot/.local/bin/redbot-setup", line 8, in <module>
    sys.exit(run_cli())
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/setup.py", line 461, in run_cli
    cli()  # pylint: disable=no-value-for-parameter  # click
  File "/home/redbot/.local/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/redbot/.local/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/redbot/.local/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/redbot/.local/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/redbot/.local/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/setup.py", line 455, in backup
    asyncio.run(create_backup(instance, Path(destination_folder)))
  File "/usr/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/setup.py", line 256, in create_backup
    backup_fpath = await red_create_backup(destination_folder)
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/core/utils/_internal_utils.py", line 233, in create_backup
    await repo_mgr.initialize()
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/cogs/downloader/repo_manager.py", line 1013, in initialize
    await self._load_repos(set_repos=True)
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/cogs/downloader/repo_manager.py", line 1208, in _load_repos
    branch = await self.config.repos.get_raw(folder.stem, default="")
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/core/config.py", line 588, in get_raw
    raw = await self.driver.get(identifier_data)
  File "/home/redbot/.local/lib/python3.8/site-packages/redbot/core/drivers/redis/redis.py", line 196, in get
    if not await self._pool.exists(cog_name):
  File "/home/redbot/.local/lib/python3.8/site-packages/aioredis/pool.py", line 257, in _wait_execute
    conn = await self.acquire(command, args)
  File "/home/redbot/.local/lib/python3.8/site-packages/aioredis/pool.py", line 319, in acquire
    raise PoolClosedError("Pool is closed")
aioredis.errors.PoolClosedError: Pool is closed
redbot@laggron-VPS:~$
```

**Pull request closed without merging**